sympy/algebras/quaternion.py

```diff
@@ -84,6 +111,159 @@ def d(self):
+
+        # calculate distinct quaternions
+        qi = cls.from_axis_angle(ei, angles[0])
+        qj = cls.from_axis_angle(ej, angles[1])
+        qk = cls.from_axis_angle(ek, angles[2])
+
+        if extrinsic:
+            return trigsimp(qk * qj * qi)
+        else:
+            return trigsimp(qi * qj * qk)
+
+    def to_euler(self, seq):
+        """Returns Euler angles representing same in the sequence given by
+        `seq`. This implements the method described in [1]_.
+
+        Parameters
+        ==========
+
+        seq : string of length 3
+            Represents the sequence of rotations.
+            For intrinsic rotations, seq but be all lowercase and its elements
+            must be from the set `['x', 'y', 'z']`
+            For extrinsic rotations, seq but be all uppercase and its elements
+            must be from the set `['X', 'Y', 'Z']`
+
+        Returns
+        =======
+
+        Matrix
+            The Euler angles calculated from the quaternion
+
+        Examples
+        ========
+
+        >>> from sympy import Quaternion
+        >>> from sympy.abc import a, b, c, d
+        >>> euler = Quaternion(a, b, c, d).to_euler('zyz')
+        >>> euler
+        Matrix([
+            [                  -atan2(-b, c) + atan2(d, a)],
+            [2*atan2(sqrt(b**2 + c**2), sqrt(a**2 + d**2))],
+            [                   atan2(-b, c) + atan2(d, a)]])
+
+
+        References
+        ==========
+
+        .. [1] https://doi.org/10.1371/journal.pone.0276302
+
+        """
+        extrinsic = _check_sequence(seq)
+        i, j, k = seq.lower()
+
+        # get index corresponding to elementary basis vectors
+        i = 'xyz'.index(i) + 1
+        j = 'xyz'.index(j) + 1
+        k = 'xyz'.index(k) + 1
+
+        if not extrinsic:
+            i, k = k, i
+
+        # check if sequence is symmetric
+        symmetric = i == k
+        if symmetric:
+            k = 6 - i - j
+
+        # parity of the permutation
+        sign = (i - j) * (j - k) * (k - i) // 2
+
+        # permutate elements
+        elements = [self.a, self.b, self.c, self.d]
+        a = elements[0]
+        b = elements[i]
+        c = elements[j]
+        d = elements[k] * sign
+
+        if not symmetric:
+            a, b, c, d = a - c, b + d, c + a, d - b
+
+        # calculate angles
+        half_sum = atan2(b, a)
+        half_diff = atan2(d, c)
+
+        angle_j = 2*atan2(sqrt(c*c + d*d), sqrt(a*a + b*b))
+        # alternatively, we can use this to avoid the square root:
+        # angle_2 = acos(2*(a*a + b*b)/(a*a + b*b + c*c + d*d) - 1)
+
+        angle_i = half_sum + half_diff
+        angle_k = half_sum - half_diff
+
+        # for Tait-Bryan angles
+        if not symmetric:
+            angle_j -= pi / 2
+            angle_i *= sign
+
+        if extrinsic:
+            return Matrix([angle_k, angle_j, angle_i])
+        else:
+            return Matrix([angle_i, angle_j, angle_k])
```

Comment:
Do we need `Matrix` to store this result? 
I think that if it is well defined positional arguments, we can pick the simplest representation like `tuple`

Reply:
Not at all, you're right. I'll change it to a tuple on my next modifications.

Reply:
Done!